# Heart Failure

In [1]:
import pandas as pd

## Retrieve the Data

**The Data was retrieved from Kaggle : [https://www.kaggle.com/datasets/fedesoriano/heart-failure-prediction]**

In [2]:
# Import the data into a dataframe
path = "resources/heart.csv"
data = pd.read_csv(path)

data.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [3]:
# Check all the Fields
data.columns

Index(['Age', 'Sex', 'ChestPainType', 'RestingBP', 'Cholesterol', 'FastingBS',
       'RestingECG', 'MaxHR', 'ExerciseAngina', 'Oldpeak', 'ST_Slope',
       'HeartDisease'],
      dtype='object')

In [4]:
data.describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR  \
count  918.000000  918.000000   918.000000  918.000000  918.000000   
mean    53.510893  132.396514   198.799564    0.233115  136.809368   
std      9.432617   18.514154   109.384145    0.423046   25.460334   
min     28.000000    0.000000     0.000000    0.000000   60.000000   
25%     47.000000  120.000000   173.250000    0.000000  120.000000   
50%     54.000000  130.000000   223.000000    0.000000  138.000000   
75%     60.000000  140.000000   267.000000    0.000000  156.000000   
max     77.000000  200.000000   603.000000    1.000000  202.000000   

          Oldpeak  HeartDisease  
count  918.000000    918.000000  
mean     0.887364      0.553377  
std      1.066570      0.497414  
min     -2.600000      0.000000  
25%      0.000000      0.000000  
50%      0.600000      1.000000  
75%      1.500000      1.000000  
max      6.200000      1.000000

1. Age Range 28 - 77 years old

In [5]:
data['Sex'].value_counts()

Sex
M    725
F    193
Name: count, dtype: int64

In [6]:
data['HeartDisease'].value_counts()

HeartDisease
1    508
0    410
Name: count, dtype: int64

This data shows that 508  people has the heart disease out of 4

In [7]:
# Confirm all the values exists
data.isnull().mean()

Age               0.0
Sex               0.0
ChestPainType     0.0
RestingBP         0.0
Cholesterol       0.0
FastingBS         0.0
RestingECG        0.0
MaxHR             0.0
ExerciseAngina    0.0
Oldpeak           0.0
ST_Slope          0.0
HeartDisease      0.0
dtype: float64

In [8]:
numerical_fields = ['Age', 'RestingBP', 'Cholesterol', 'FastingBS','MaxHR', 'Oldpeak']

categorical_fields = ['Sex', 'ChestPainType', 'RestingECG', 'ExerciseAngina', 'ST_Slope']

In [9]:
data[numerical_fields].describe()

Age   RestingBP  Cholesterol   FastingBS       MaxHR     Oldpeak
count  918.000000  918.000000   918.000000  918.000000  918.000000  918.000000
mean    53.510893  132.396514   198.799564    0.233115  136.809368    0.887364
std      9.432617   18.514154   109.384145    0.423046   25.460334    1.066570
min     28.000000    0.000000     0.000000    0.000000   60.000000   -2.600000
25%     47.000000  120.000000   173.250000    0.000000  120.000000    0.000000
50%     54.000000  130.000000   223.000000    0.000000  138.000000    0.600000
75%     60.000000  140.000000   267.000000    0.000000  156.000000    1.500000
max     77.000000  200.000000   603.000000    1.000000  202.000000    6.200000

In [10]:
# Preview the category columns in the raw format
# ---------------------------------------------

# Before one-hot-encoded
before_one_hot_encoded =data.copy()

before_one_hot_encoded = before_one_hot_encoded.drop(columns=numerical_fields)
del(before_one_hot_encoded['HeartDisease'])

before_one_hot_encoded

Sex ChestPainType RestingECG ExerciseAngina ST_Slope
0     M           ATA     Normal              N       Up
1     F           NAP     Normal              N     Flat
2     M           ATA         ST              N       Up
3     F           ASY     Normal              Y     Flat
4     M           NAP     Normal              N       Up
..   ..           ...        ...            ...      ...
913   M            TA     Normal              N     Flat
914   M           ASY     Normal              N     Flat
915   M           ASY     Normal              Y     Flat
916   F           ATA        LVH              N     Flat
917   M           NAP     Normal              N       Up

[918 rows x 5 columns]

In [11]:
# OneHotEncoder on columns with categorical data
from sklearn.preprocessing import OneHotEncoder

# Create an instance of OneHotEncoder()
enc = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

# Columns to one hot encode
columns_to_enocde = categorical_fields

# Fit the encoder to the data
# Identifies the unique categories of each column and assigns a binary vector 
enc.fit(data[columns_to_enocde])

# Transform the data
# Each unique category becomes a separate binary column
data_one_hot_encoded = enc.transform(data[columns_to_enocde])


# Create a pandas dataframe to join this with the original dataframe
data_one_hot_encoded_df = pd.DataFrame(
    data_one_hot_encoded,
    columns = enc.get_feature_names_out(columns_to_enocde),
    index= data.index
)


In [12]:
print(type(data))
print(type(data_one_hot_encoded_df))

<class 'pandas.core.frame.DataFrame'>
<class 'pandas.core.frame.DataFrame'>


In [13]:
# Preview the one-hot-encoded dataframe
data_one_hot_encoded_df

Sex_F  Sex_M  ChestPainType_ASY  ChestPainType_ATA  ChestPainType_NAP  \
0      0.0    1.0                0.0                1.0                0.0   
1      1.0    0.0                0.0                0.0                1.0   
2      0.0    1.0                0.0                1.0                0.0   
3      1.0    0.0                1.0                0.0                0.0   
4      0.0    1.0                0.0                0.0                1.0   
..     ...    ...                ...                ...                ...   
913    0.0    1.0                0.0                0.0                0.0   
914    0.0    1.0                1.0                0.0                0.0   
915    0.0    1.0                1.0                0.0                0.0   
916    1.0    0.0                0.0                1.0                0.0   
917    0.0    1.0                0.0                0.0                1.0   

     ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  RestingECG_ST  \
0                 0.0             0.0                1.0            0.0   
1                 0.0             0.0                1.0            0.0   
2                 0.0             0.0                0.0            1.0   
3                 0.0             0.0                1.0            0.0   
4                 0.0             0.0                1.0            0.0   
..                ...             ...                ...            ...   
913               1.0             0.0                1.0            0.0   
914               0.0             0.0                1.0            0.0   
915               0.0             0.0                1.0            0.0   
916               0.0             1.0                0.0            0.0   
917               0.0             0.0                1.0            0.0   

     ExerciseAngina_N  ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  \
0                 1.0               0.0            0.0            0.0   
1                 1.0               0.0            0.0            1.0   
2                 1.0               0.0            0.0            0.0   
3                 0.0               1.0            0.0            1.0   
4                 1.0               0.0            0.0            0.0   
..                ...               ...            ...            ...   
913               1.0               0.0            0.0            1.0   
914               1.0               0.0            0.0            1.0   
915               0.0               1.0            0.0            1.0   
916               1.0               0.0            0.0            1.0   
917               1.0               0.0            0.0            0.0   

     ST_Slope_Up  
0            1.0  
1            0.0  
2            1.0  
3            0.0  
4            1.0  
..           ...  
913          0.0  
914          0.0  
915          0.0  
916          0.0  
917          1.0  

[918 rows x 14 columns]

In [14]:
# join the original df with the one-hot encoded binary columns
data = pd.concat([data, data_one_hot_encoded_df], axis=1)

# delete the original categorical fields
data = data.drop(columns=categorical_fields)
data

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0     40        140          289          0    172      0.0             0   
1     49        160          180          0    156      1.0             1   
2     37        130          283          0     98      0.0             0   
3     48        138          214          0    108      1.5             1   
4     54        150          195          0    122      0.0             0   
..   ...        ...          ...        ...    ...      ...           ...   
913   45        110          264          0    132      1.2             1   
914   68        144          193          1    141      3.4             1   
915   57        130          131          0    115      1.2             1   
916   57        130          236          0    174      0.0             1   
917   38        138          175          0    173      0.0             0   

     Sex_F  Sex_M  ChestPainType_ASY  ...  ChestPainType_NAP  \
0      0.0    1.0                0.0  ...                0.0   
1      1.0    0.0                0.0  ...                1.0   
2      0.0    1.0                0.0  ...                0.0   
3      1.0    0.0                1.0  ...                0.0   
4      0.0    1.0                0.0  ...                1.0   
..     ...    ...                ...  ...                ...   
913    0.0    1.0                0.0  ...                0.0   
914    0.0    1.0                1.0  ...                0.0   
915    0.0    1.0                1.0  ...                0.0   
916    1.0    0.0                0.0  ...                0.0   
917    0.0    1.0                0.0  ...                1.0   

     ChestPainType_TA  RestingECG_LVH  RestingECG_Normal  RestingECG_ST  \
0                 0.0             0.0                1.0            0.0   
1                 0.0             0.0                1.0            0.0   
2                 0.0             0.0                0.0            1.0   
3                 0.0             0.0                1.0            0.0   
4                 0.0             0.0                1.0            0.0   
..                ...             ...                ...            ...   
913               1.0             0.0                1.0            0.0   
914               0.0             0.0                1.0            0.0   
915               0.0             0.0                1.0            0.0   
916               0.0             1.0                0.0            0.0   
917               0.0             0.0                1.0            0.0   

     ExerciseAngina_N  ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  \
0                 1.0               0.0            0.0            0.0   
1                 1.0               0.0            0.0            1.0   
2                 1.0               0.0            0.0            0.0   
3                 0.0               1.0            0.0            1.0   
4                 1.0               0.0            0.0            0.0   
..                ...               ...            ...            ...   
913               1.0               0.0            0.0            1.0   
914               1.0               0.0            0.0            1.0   
915               0.0               1.0            0.0            1.0   
916               1.0               0.0            0.0            1.0   
917               1.0               0.0            0.0            0.0   

     ST_Slope_Up  
0            1.0  
1            0.0  
2            1.0  
3            0.0  
4            1.0  
..           ...  
913          0.0  
914          0.0  
915          0.0  
916          0.0  
917          1.0  

[918 rows x 21 columns]

## Splitting the data

In [15]:
# Import module
from sklearn.model_selection import train_test_split

# Split training and testing sets
# Create the features Dataframe X
X = data.copy()
X = data.drop(columns = ['HeartDisease'])

# Create the target Dataframe, y
y = data['HeartDisease']

# Use train_test_split to separate the data
X_train, X_test, y_train, y_test = train_test_split(X, y)



### Apply Standard Scaler

In [16]:
# StandardScaler on columns with numerical data
from sklearn.preprocessing import StandardScaler

# Scale the x data by using Standard Scaler
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_train_scaled

array([[ 1.44851399, -1.72764608,  0.93908798, ..., -0.27691572,
         1.04459821, -0.90559694],
       [-1.09053729, -0.56181233, -1.81836536, ..., -0.27691572,
        -0.95730588,  1.10424401],
       [ 0.49636976, -0.03188789,  0.24741908, ..., -0.27691572,
        -0.95730588,  1.10424401],
       ...,
       [ 0.28478216,  1.1869383 , -1.81836536, ..., -0.27691572,
         1.04459821, -0.90559694],
       [-0.13839306, -0.7737821 , -0.10302649, ..., -0.27691572,
         1.04459821, -0.90559694],
       [ 0.39057596, -1.46268386, -1.81836536, ..., -0.27691572,
         1.04459821, -0.90559694]])

In [17]:
# Scale the x  test data using Standard scaler
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-1.30212489, -1.46268386,  0.00764053, ..., -0.27691572,
        -0.95730588,  1.10424401],
       [-0.03259925,  0.39205165,  0.17364106, ..., -0.27691572,
         1.04459821, -0.90559694],
       [ 1.02533878,  2.7767116 , -1.81836536, ..., -0.27691572,
        -0.95730588,  1.10424401],
       ...,
       [-2.25426912, -1.19772165,  0.25664133, ..., -0.27691572,
        -0.95730588,  1.10424401],
       [ 0.91954498, -0.66779721,  0.52408664, ..., -0.27691572,
         1.04459821, -0.90559694],
       [ 1.6601016 ,  0.49803654,  0.68086492, ..., -0.27691572,
        -0.95730588,  1.10424401]])

In [18]:
 # Check the max and min of the scaled training and testing sets
print("Scaled data min/max (StandardScaler):")
print("Training data min:",X_train_scaled.min())
print("Training data max:",X_train_scaled.max())
print("Testing data min:",X_test_scaled.min())
print("Testing data max:",X_test_scaled.max())


Scaled data min/max (StandardScaler):
Training data min: -7.02689041519964
Training data max: 5.136211665136863
Testing data min: -3.3245490426513022
Testing data max: 4.559341616878578


### Model and Fit the Data to a Logistic Regression

In [28]:
# Import 'LogisticRegression' from sklean
from sklearn.linear_model import LogisticRegression

# Create an instance of LogisticRegression
logistic_regression_model = LogisticRegression()

# Fit the model
logistic_regression_model.fit(X_train_scaled, y_train)



Training Data Score: 0.87645


In [31]:
# Predict the test data
lr_predictions = logistic_regression_model.predict(X_test_scaled)

lr_predictions

array([0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1,
       1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1,
       1, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0, 0, 0,
       1, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 0, 1, 0, 1, 1, 1, 0, 1, 0], dtype=int64)

In [34]:
# Save both the test predictions and actual test values to a DataFrame
results_df = pd.DataFrame({
    "Testing Data Predictions": lr_predictions, 
    "Testing Data Actual Targets": y_test})

results_df

Testing Data Predictions  Testing Data Actual Targets
670                         0                            0
87                          0                            0
372                         1                            1
248                         1                            1
249                         1                            1
..                        ...                          ...
358                         1                            1
663                         1                            1
67                          0                            0
614                         1                            1
583                         0                            0

[230 rows x 2 columns]

In [49]:
# Import the accuracy_score function
from sklearn.metrics import accuracy_score

lr_train_score = logistic_regression_model.score(X_train_scaled, y_train)
lr_test_score = accuracy_score(y_test, lr_predictions)
score_gap = lr_train_score - lr_test_score

# Calculate the model's accuracy on the test dataset
print(f"Training Data Score: {lr_train_score}")
print(f"Testing Data Score: {lr_test_score}")
print(f"Score gap: {score_gap}")

Training Data Score: 0.876453488372093
Testing Data Score: 0.8434782608695652
Score gap: 0.03297522750252779
